pkg/gitreceive/build.go

```diff
@@ -0,0 +1,295 @@
+// repoCmd returns exec.Command(first, others...) with its current working directory repoDir
+func repoCmd(repoDir, first string, others ...string) *exec.Cmd {
+	cmd := exec.Command(first, others...)
+	cmd.Dir = repoDir
+	return cmd
+}
+
+// mcCmd returns a command to execute the 'mc' binary, so that it reads config from configDir.
+// the command outputs its stderr to os.Stderr
+func mcCmd(configDir string, args ...string) *exec.Cmd {
+	cmd := exec.Command("mc", "-C", configDir, "--quiet")
+	cmd.Args = append(cmd.Args, args...)
+	cmd.Stderr = os.Stderr
+	return cmd
+}
+
+func kGetCmd(podNS, podName string) *exec.Cmd {
+	return exec.Command("kubectl", fmt.Sprintf("--namespace=%s", podNS), "get", "pods", "-o", "yaml", podName)
+}
+
+// run prints the command it will execute to the debug log, then runs it and returns the result of run
+func run(cmd *exec.Cmd) error {
+	cmdStr := strings.Join(cmd.Args, " ")
+	if cmd.Dir != "" {
+		log.Debug("running [%s] in directory %s", cmdStr, cmd.Dir)
+	} else {
+		log.Debug("running [%s]", cmdStr)
+	}
+	return cmd.Run()
+}
+
+func build(conf *Config, builderKey, gitSha string) error {
+	storage, err := getStorageConfig()
+	if err != nil {
+		return err
+	}
+	creds, err := getStorageCreds()
+	if err == errMissingKey || err == errMissingSecret {
+		return err
+	}
+
+	repo := conf.Repository
+	if len(gitSha) <= shortShaIdx {
+		return errGitShaTooShort{sha: gitSha}
+	}
+	shortSha := gitSha[0:8]
+	appName := conf.App()
+
+	repoDir := filepath.Join(conf.GitHome, repo)
+	buildDir := filepath.Join(repoDir, "build")
+
+	slugName := fmt.Sprintf("%s:git-%s", appName, shortSha)
+	imageName := strings.Replace(slugName, ":", "-", -1)
+	if err := os.MkdirAll(buildDir, os.ModeDir); err != nil {
+		return fmt.Errorf("making the build directory %s (%s)", buildDir, err)
+	}
+	tmpDir := os.TempDir()
+
+	tarURL := fmt.Sprintf("%s://%s:%s/git/home/%s/tar", storage.schema(), storage.host(), storage.port(), slugName)
+
+	// this is where workflow tells slugrunner to download the slug from, so we have to tell slugbuilder to upload it to here
+	pushURL := fmt.Sprintf("%s://%s:%s/git/home/%s/push", storage.schema(), storage.host(), storage.port(), fmt.Sprintf("%s:git-%s", appName, gitSha))
+
+	// Ensure that the app config can be gotten from workflow. We don't do anything with this information
+	appConf, err := getAppConfig(conf, builderKey, conf.Username, appName)
+	if err != nil {
+		return fmt.Errorf("getting app config for %s (%s)", appName, err)
+	}
+	log.Debug("got the following config back for app %s: %+v", appName, *appConf)
+	var buildPackURL string
+	if buildPackURLInterface, ok := appConf.Values["BUILDPACK_URL"]; ok {
+		if bpStr, ok := buildPackURLInterface.(string); ok {
+			log.Debug("found custom buildpack URL %s", bpStr)
+			buildPackURL = bpStr
+		}
+	}
+
+	// build a tarball from the new objects
+	gitArchiveCmd := repoCmd(repoDir, "git", "archive", "--format=tar.gz", fmt.Sprintf("--output=%s.tar.gz", appName), gitSha)
```

Comment:
appname.tar.gz is being used in a couple places (line 117). Should we make a variable just for best practice?